virtual_rainforest/core/schema.py

```diff
@@ -0,0 +1,148 @@
+        json.JSONDecodeError: the file at the schema path is not valid JSON
+        jsonschema.SchemaError: the file contents are not valid JSON Schema
+        ValueError: the JSON Schema is missing required keys
+    """
+
+    # Try and get the contents of the JSON schema file
+    try:
+        json_schema = json.load(open(schema_file_path))
+    except FileNotFoundError:
+        fnf_error = FileNotFoundError(f"Schema file not found {schema_file_path}.")
+        LOGGER.error(fnf_error)
+        raise fnf_error
+    except json.JSONDecodeError as excep:
+        LOGGER.error(f"JSON error in schema file {schema_file_path}")
+        raise excep
+
+    # Check that this is a valid schema - deliberately log a separate message and let
+    # the schema traceback output rather than replacing it.
+    try:
+        ValidatorWithDefaults.check_schema(json_schema)
+    except exceptions.SchemaError as excep:
+        LOGGER.error(f"Module schema invalid in: {schema_file_path}")
+        raise excep
+
+    # Check that relevant keys are included - deliberately re-raising as ValueError here
+    # as the KeyError has built in quoting and is expecting to provide simple missing
+    # key messages.
+    try:
+        json_schema["properties"][module_name]
+        json_schema["required"]
+    except KeyError as excep:
+        to_raise = ValueError(f"Missing key in module schema {module_name}: {excep}")
+        LOGGER.error(to_raise)
+        raise to_raise
+
+    return json_schema
+
+
+def merge_schemas(schemas: dict[str, dict[str, Any]]) -> dict[str, Any]:
+    """Merge the validation schemas for desired modules.
+
+    The method merges a set of schemas for a set of desired modules into a single
+    integrated schema that can then be used to validate a merged configuration for those
+    modules.
+
+    Args:
+        schema: A dictionary of schemas keyed by module name
+
+    Returns:
+        An integrated schema combining the modules.
+    """
+
+    # Construct combined schema for all relevant modules
+    comb_schema: dict = {"type": "object", "properties": {}, "required": []}
+
+    # Loop over expected modules amd add them to the combined schema
+    for this_module, this_schema in schemas.items():
+        comb_schema["properties"][this_module] = this_schema["properties"][this_module]
+        # Add module name to list of required modules
+        comb_schema["required"].append(this_module)
+
+    p_paths = []
+    # Recursively search for all instances of properties in the schema
+    for path, value in dpath.util.search(comb_schema, "**/properties", yielded=True):
+        # Remove final properties instance from path so that additionalProperties ends
+        # up in the right place
+        p_paths.append("" if path == "properties" else path[:-11])
+
+    # Set additional properties to false everywhere that properties are defined
+    for path in p_paths:
+        dpath.util.new(comb_schema, f"{path}/additionalProperties", False)
```

Comment:
Could these two loops be combined? Then you could drop `p_paths`

Reply:
Merging the loops results in the code trying to edit the dictionary whilst it is being iterated over, and it objects. But it can be expressed probably more cleanly by generating the list first and then looping over doing the edits.

Reply:
I think that will be much neater, to be honest.